### PR TITLE
feat: add support to collect metrics using the OTEL library

### LINKED ISSUE #910: Interceptor Proxy - OTEL Instrumentation
### Proposal

Currently the interceptor proxy does not emit any metrics or traces which makes it difficult to understand how the service is performing. It should use OTEL instrumentation to provide observability, making it easier to monitor key performance attributes of the service in a real world setting.

### Use-Case

Running the interceptor proxy in a production setting where having insight into the performance of your platform services is key.

### Is this a feature you are interested in implementing yourself?

Yes

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by zorocloud**
Adding support to the interceptor proxy to allow for the collection of metrics via a prometheus compatible endpoint or via a OTEL HTTP collector. Resolves #910 

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] Any necessary documentation is added, such as:
  - [`README.md`](/README.md)
  - [The `docs/` directory](./docs)
  - [The docs repo](https://github.com/kedacore/keda-docs)

### CODE REVIEW

**Review by JorTurFer — COMMENTED** [review]
Looks really nice! Could you add some documentation about how to enable these metrics? I think that we could create a new section like `Operate` and include this information about how to enable prometheus/otel and which metrics are available.

do you think that adding an e2e test could be useful? In KEDA core we cover the metrics too using e2e tests, so we could just add an e2e test like them

**Comment by zorocloud** [comment]
> Looks really nice! Could you add some documentation about how to enable these metrics? I think that we could create a new section like `Operate` and include this information about how to enable prometheus/otel and which metrics are available.
Thanks! Yeah, sure. Will draft some docs up and get the PR updated.

> do you think that adding an e2e test could be useful? In KEDA core we cover the metrics too using e2e tests, so we could just add an e2e test like them
That is a good shout. Will get a test for both cases added in and update the PR.

**Comment by zorocloud** [comment]
I have added a brief overview of how to configure each exporter in a Operate.md doc as per your suggestion @JorTurFer and added in some e2e tests (using the core Keda set as a reference). Let me know your thoughts :)

File: config/interceptor/deployment.yaml
Comment (zorocloud) [reply config/interceptor/deployment.yaml:69]: It's the one they used in the OTEL example, just kept it the same tbh lol

File: tests/checks/interceptor_prometheus_metrics/interceptor_prometheus_metrics_test.go
Comment (JorTurFer) [inline tests/checks/interceptor_prometheus_metrics/interceptor_prometheus_metrics_test.go:173]: I'd assert this using greater or equal because the Prometheus metrics are persisted over the time, and 2nd and 3rd tries can fail because the metric returns 2,3, etc

File: tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go
Comment (JorTurFer) [inline tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go:154]: This e2e test has failed due to compiling issues

File: tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go
Comment (JorTurFer) [inline tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go:182]: I'd assert this using greater or equal because the Prometheus metrics are persisted over the time, and 2nd and 3rd tries can fail because the metric returns 2,3, etc

File: tests/checks/interceptor_prometheus_metrics/interceptor_prometheus_metrics_test.go
Comment (zorocloud) [reply tests/checks/interceptor_prometheus_metrics/interceptor_prometheus_metrics_test.go]: Yeah that is a good shout - when I had tested I was just tearing it down after every run so only ever had a count of 1 but that won't always be true

File: tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go
Comment (zorocloud) [reply tests/checks/interceptor_otel_metrics/interceptor_otel_metrics_test.go:153]: Oops, left a small typo in there. Removed.

**Comment by zorocloud** [comment]
Will take a look why the e2e tests failed when I get a sec

**Review by JorTurFer — APPROVED** [review]
LGTM!

**Comment by JorTurFer** [comment]
There are some static code errors, could you take a look please? https://github.com/kedacore/http-add-on/actions/runs/8453796744/job/23280002430?pr=927

File: docs/operate.md
Comment (tomkerkhove) [inline docs/operate.md:10]:
```suggestion
When configured, the interceptor proxy can expose metrics on a Prometheus compatible endpoint.

This endpoint can be enabled by setting the `KEDA_HTTP_OTEL_PROM_EXPORTER_ENABLED` environment variable to `true` on the interceptor deployment (`true` by default) and by setting `KEDA_HTTP_OTEL_PROM_EXPORTER_PORT` to an unused port for the endpoint to be made available on (`2223` by default).
```

File: docs/operate.md
Comment (tomkerkhove) [inline docs/operate.md:13]:
```suggestion
When configured, the interceptor proxy can export metrics to a OTEL HTTP collector.

The OTEL exporter can be enabled by setting the `KEDA_HTTP_OTEL_HTTP_EXPORTER_ENABLED` environment variable to `true` on the interceptor deployment (`false` by default). When enabled the `KEDA_HTTP_OTEL_HTTP_COLLECTOR_ENDPOINT` environment variable must also be configured so the exporter knows what collector to send the metrics to (e.g. opentelemetry-collector.open-telemetry-system:4318).

If the collector is exposed on an unsecured endpoint then you can set the `KEDA_HTTP_OTEL_HTTP_COLLECTOR_INSECURE` environment variable to `true` (`false` by default) which will disable client security on the exporter.

If you need to provide any headers such as authentication details in order to utilise your OTEL collector you can add them into the `KEDA_HTTP_OTEL_HTTP_HEADERS` environment variable. The frequency at which the metrics are exported can be configured by setting `KEDA_HTTP_OTEL_METRIC_EXPORT_INTERVAL` to the number of seconds you require between each export interval (`30` by default).
```

**Review by tomkerkhove — COMMENTED** [review]
Docs LGTM, nice work!

**Comment by zorocloud** [comment]
Thanks for the feedback. Remaining suggestions should be addressed and the static checks fixed. Just noticed I need to fix some conflicts too. Will sort that when I next get a sec :+1:

File: docs/operate.md
Comment (zorocloud) [reply docs/operate.md]: updated :)

File: docs/operate.md
Comment (zorocloud) [reply docs/operate.md]: Updated :)

**Comment by zorocloud** [comment]
I have just rebased onto main and the unit/e2e tests look like they are passing fine so we should be good now 👍

**Comment by JorTurFer** [comment]
We can ignore Synk, it has transient failures from time to time xD I have marked it as okey manually :)

**Comment by JorTurFer** [comment]
There is other PR ongoing and once that's merged, I'll update your PR and merge it. Thanks for your AWESOME contribution ❤️

File: tests/utils/setup_test.go
Comment (JorTurFer) [inline tests/utils/setup_test.go:21]:
```suggestion
image:
  repository: "otel/opentelemetry-collector-contrib"
config:
  exporters:
```

Comment (JorTurFer) [reply tests/utils/setup_test.go:21]: There is a potential breaking change that requires the change above, please commit it 🙏 
https://github.com/open-telemetry/opentelemetry-helm-charts/blob/main/charts/opentelemetry-collector/UPGRADING.md#0880-to-0890

Comment (zorocloud) [reply tests/utils/setup_test.go:23]: Done @JorTurFer :) 

Also, I've pushed a change to move the pending request count from a Gauge to a UpDownCounter (noticed some memory issues with the Gauge). Left it in a separate commit rather than squashing locally so it's easier for you to review.

**Comment by JorTurFer** [comment]
There are merge conflicts to be solved, could you solve them? (or enabling me to update your PR xD)

**Comment by zorocloud** [comment]
Should be sorted now @JorTurFer  :) 

Also, allowed maintainers to edit the PR, not sure why I didn't select this previously.

**Review by JorTurFer — APPROVED** [review]
LGTM! ❤️

**Comment by JorTurFer** [comment]
I didn't noticed during the review that we aren't respected the OTEL spec ENV vars: https://opentelemetry.io/docs/specs/otel/protocol/exporter/

I've created an issue to track it, as I guess that you are using the metrics, I share it with you because we will have to change them for future versions: https://github.com/kedacore/http-add-on/issues/1031

**Comment by zorocloud** [comment]
Ah sugar, I completely missed that too :facepalm:. I'm happy to get a PR raised toward the end of next week to fix this if someone isn't already working on it.

**Comment by JorTurFer** [comment]
>I'm happy to get a PR raised toward the end of next week to fix this if someone isn't already working on it.

I'd be nice, I just pinged you to make you aware about a possible breaking change, but if you are willing to update the code, it's awesome :)

**Pull request merged**
